crates/engine/tree/docs/root.md

```diff
@@ -0,0 +1,247 @@
+to signal that the proof calculation is finished.
+
+Some proofs can arrive earlier than others, even though they were requested later. It depends on the number
+of proof targets, and also some non-determinism in the database caching.
+
+The issue with this is that we need to ensure that the proofs are sent
+to the [Sparse Trie Task](#sparse-trie-task) in the order that they were requested. Because of this,
+we introduced a `ProofSequencer` that we add new proofs to.
+https://github.com/paradigmxyz/reth/blob/2ba54bf1c1f38c7173838f37027315a09287c20a/crates/engine/tree/src/tree/root.rs#L666-L672
+
+`ProofSequencer` acts in the following way:
+1. Each proof has an associated "sequence number" that determines the original order of state updates.
+2. When the proof is calculatd, it's added to the `ProofSequencer` with the sequence number
+and state update associated with it.
+3. If the `ProofSequencer` has a consecutive sequence of proofs without gaps in sequence numbers, it returns this sequence.
+
+Once the `ProofSequencer` returns a sequence of proofs,
+we send them along with the state updates to the [Sparse Trie Task](#sparse-trie-task).
+
+### Finishing the calculation
+
+Once all transactions are executed, the [Engine](#engine) sends a `StateRootMessage::FinishStateUpdates` message
+to the State Root Task, marking the end of receiving state updates.
+
+Every time we receive a new proof from the [MultiProof Manager](#multiproof-manager), we also check
+the following conditions:
+1. Are all updates received? (`StateRootMessage::FinishStateUpdates` was sent)
+2. Is `ProofSequencer` empty? (no proofs are pending for sequencing)
+3. Are all proofs that were sent to the [`MultiProofManager::spawn_or_queue`](#multiproof-manager) finished
+calculating and were sent to the [Sparse Trie Task](#sparse-trie-task)?
+
+https://github.com/paradigmxyz/reth/blob/2ba54bf1c1f38c7173838f37027315a09287c20a/crates/engine/tree/src/tree/root.rs#L935-L944
+
+When all conditions are met, we close the [State Root Task](#state-root-task) receiver channel,
+signaling that no proofs or state updates are coming anymore, and the state root calculation should be finished.
+
+
+## MultiProof Manager
+
+![MultiProof Manager](./mermaid/multiproof-manager.mmd.svg)
+
+MultiProof manager is a component responsible for generating MPT proofs
+and sending them back to the [State Root Task](#state-root-task).
+
+### Spawning new proof calculations
+
+The entrypoint is the `spawn_or_queue` method
+https://github.com/paradigmxyz/reth/blob/2ba54bf1c1f38c7173838f37027315a09287c20a/crates/engine/tree/src/tree/root.rs#L355-L357
+
+It has the following responsibilities:
+1. On empty proof targets, immediately send `StateRootMessage::EmptyProof` to the [State Root Task](#state-root-task).
+2. If the number of maximum concurrent proof calculations is reached, push the proof request to the pending queue.
```

Comment:
what do we base the max num of concurrent proofs on ?

Reply:
clarified how we calculate it